MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs

```diff
@@ -80,66 +69,61 @@ private async void UiElementOnMouseLeave(object sender, MouseEventArgs mouseEven
                 lock (_waitHandleGate)
                 {
-                    if (!_isWaitHandleDisposed)
+                    if (!_isDisposed)
                         _waitHandle.Set();
                 }
             }
 
-            private void UiElementOnMouseEnter(object sender, MouseEventArgs mouseEventArgs) 
-                => _ = _waitHandle.Reset();
+            public void Dispose()
+            {
+                if (_isDisposed)
+                    return;
 
-            public void Dispose() => _cleanUp();
+                _uiElement.MouseEnter -= UiElementOnMouseEnter;
+                _uiElement.MouseLeave -= UiElementOnMouseLeave;
+                lock (_waitHandleGate)
+                {
+                    _waitHandle.Dispose();
+                    _isDisposed = true;
+                }
+                _disposedWaitHandle.Set();
+                _disposedWaitHandle.Dispose();
+            }
         }
 
         #endregion
 
         #region private class DurationMonitor
 
-        private class DurationMonitor
+        private static class DurationMonitor
         {
-            private DateTime _completionTime;
-
-            private DurationMonitor(
-                TimeSpan minimumDuration,
-                WaitHandle pausedWaitHandle,
-                EventWaitHandle signalWhenDurationPassedWaitHandle,
-                WaitHandle ceaseWaitHandle)
+            public static void Start(TimeSpan minimumDuration, WaitHandle pausedWaitHandle, EventWaitHandle signalWhenDurationPassedWaitHandle, WaitHandle ceaseWaitHandle)
             {
                 if (pausedWaitHandle == null) throw new ArgumentNullException(nameof(pausedWaitHandle));
-                if (signalWhenDurationPassedWaitHandle == null)
-                    throw new ArgumentNullException(nameof(signalWhenDurationPassedWaitHandle));
+                if (signalWhenDurationPassedWaitHandle == null) throw new ArgumentNullException(nameof(signalWhenDurationPassedWaitHandle));
                 if (ceaseWaitHandle == null) throw new ArgumentNullException(nameof(ceaseWaitHandle));
 
-                _completionTime = DateTime.Now.Add(minimumDuration);
+                var completionTime = DateTime.Now.Add(minimumDuration);
 
                 //this keeps the event waiting simpler, rather that actually watching play -> pause -> play -> pause etc
                 var granularity = TimeSpan.FromMilliseconds(200);
 
                 Task.Factory.StartNew(() =>
                 {
                     //keep upping the completion time in case it's paused...
-                    while (DateTime.Now < _completionTime && !ceaseWaitHandle.WaitOne(granularity)
-                        && !signalWhenDurationPassedWaitHandle.WaitOne(TimeSpan.Zero))
+                    while (DateTime.Now < completionTime && !ceaseWaitHandle.WaitOne(granularity)
+                                                          && !signalWhenDurationPassedWaitHandle.WaitOne(TimeSpan.Zero))
```

Comment:
Maybe extract that condition with a fitting name?

Reply:
I updated this method